### PR TITLE
Drop the use of the deprecated setuptools Features

### PR DESCRIPTION
**Pull request by jamadden**
https://github.com/pypa/setuptools/issues/65

### CODE REVIEW

File: setup.py [hunk 1]
```diff
@@ -64,24 +64,19 @@ def read(*rnames):
         return f.read()
 
 
-Cwrapper = Feature(
-    "C wrapper",
-    standard=True,
-    headers=[os.path.join('src', 'zope', 'proxy', 'proxy.h')],
```
Comment (mgedmin) [inline setup.py]: I may be overthinking this.  I think all we need to do is define a `headers` variable that's either an empty list or this list, and then pass it to `setup(..., headers=headers, ext_modules=ext_modules, ...)`.

File: setup.py [hunk 2]
```diff
@@ -64,24 +64,19 @@ def read(*rnames):
...
+        "zope.proxy._zope_proxy_proxy",
+        [os.path.join('src', 'zope', 'proxy', "_zope_proxy_proxy.c")],
+    ),
+]
 
 # PyPy won't build the extension.
 is_pypy = platform.python_implementation() == 'PyPy'
 if is_pypy:
-    features = {}
+    ext_modules = []
```
Comment (mgedmin) [inline setup.py]: Also, maybe please add an `or is_pure` check with the `os.getenv('PURE_PYTHON')` logic copied from zope.interface etc.?  So people can ask to skip extension modules.

**Comment by jamadden** [comment]
Excellent questions. Short answer, I don't know. 

In my experience with gevent, getting access to headers from other projects is not very straightforward, even when they install them correctly. A buildout, for example, does things differently than a virtual environment, and there's also issues with debug builds of Python putting things in different places. So gevent winds up with path settings for all those things, plus vendors its own fallback headers...

File: setup.py (hunk 2 above)
Comment (jamadden) [reply setup.py]: Personally, I would like to consider `PURE_PYTHON` at *build* time a deprecated mistake. Runtime, yes, let's be able to select whether to use a C extension or not, but you can't do that if it wasn't even compiled.

File: setup.py (hunk 1 above)
Comment (jamadden) [reply setup.py]: That's what [greenlet does](https://github.com/python-greenlet/greenlet/blob/829b44f62b024a3f9b79453b1c0c6a2eb766c336/setup.py#L81). I'll do that here. Thanks!

Comment (jamadden) [reply setup.py]: With `headers=` as a keyword argument, I get this path installed into my virtualenv:
`$VIRTALENV/include/site/python3.8/zope.proxy/proxy.h`, which is exactly the same path I got on `master` using the `Feature`.

**Comment by mgedmin** [comment]
so, zope.container:
- does `#include "zope.proxy/proxy.h"`
- has a `setup_requires=['zope.proxy > ...'`] 
- specifies `include_dirs=['include']` for its Extension()
- has an `./include/zope.proxy/proxy.h` in git
- bundles the ./include/zope.proxy/proxy.h in its sdist

meanwhile zope.security
- does `#include "zope/proxy/proxy.h"`
- has a `setup_requires=['zope.proxy > ...'`] 
- specifies `include_dirs=[ModuleHeaderDir('zope.proxy)]` with a custom ModuleHeaderDir class defined right there in the setup.py, with custom logic in its `__str__` (!!) that relies on pkg_resources and expects .../site-packages/zope/proxy/../../zope/proxy/proxy.h to exist.

and it looks like zope.proxy ships a zope/proxy/proxy.h via `include_pacakge_data=True` and `recursive-include src *`

I'm not sure what the conclusions are.

**Comment by jamadden** [comment]
zope.container also includes a complete copy of `_zope_proxy_proxy.c` from this repo and uses that file to extend:

```C
// src/zope/container/_zope_container_contained.c

/* Supress inclusion of the original proxy.h */
#define _proxy_H_ 1

/* Incude the proxy C source */
#include "_zope_proxy_proxy.c"
```

That includes defining the module init symbols (`init_zope_proxy_proxy`) so I'm sort of amazed that both extensions can be loaded successfully. 

```C
// src/zope/container/_zope_proxy_proxy.c
/*
 *   This file is also used as a really extensive macro in
 *   ../container/_zope_container_contained.c.  If you need to
 *   change this file, you need to "svn copy" it to ../container/.
 *
 *   This approach is taken to allow the sources for the two packages
 *   to be compilable when the relative locations of these aren't
 *   related in the same way as they are in a checkout.
 *
 *   This will be revisited in the future, but works for now.
 */
```

Maybe zope.security has it right, it doesn't have its own copy of `_zope_proxy_proxy.c`.

File: setup.py (hunk 2 above)
Comment (mgedmin) [reply setup.py]: You're right about this (pip wheel cache poisoning bad!)

**Pull request merged**